[WIP] Multi user invite link

hackerkid opened the pull request:
- [X] Implement backend
- [x] Implement basic management command
- [x] Tests for backend and management command

To test

1. Make sure to enable email invitations by opening http://localhost:9991/#organization/organization-permissions. 

2. Then run
```
./manage.py migrate
./manage.py generate_invite_links --realm zulip
```

3) Now use the generated url to sign up for new accounts.

@rishig @showell @adnrs96 I can work on the tests and remaining stuff if this approach is fine with you guys :). I think the frontend to generate the link can be a separate PR of it's own.

adnrs96 commented:
Hi @hackerkid I haven't really read out the code into details and understood the changes, just tried repeating the steps mentioned. 
I get this
```
(zulip-venv)vagrant@vagrant-base-trusty-amd64:/srv/zulip$ ./manage.py generate_invite_links --realm zulip
Traceback (most recent call last):
  File "./manage.py", line 29, in <module>
    execute_from_command_line(sys.argv)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 363, in execute_from_command_line
    utility.execute()
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 355, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/srv/zulip/zerver/management/commands/generate_invite_links.py", line 71, in handle
    url = generate_multiple_user_invite_link(realm)
  File "/srv/zulip/confirmation/models.py", line 51, in generate_multiple_user_invite_link
    InviteKey.objects.create(invite_key=key, realm=realm, date_created=timezone_now())
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/db/models/query.py", line 394, in create
    obj.save(force_insert=True, using=self.db)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/db/models/base.py", line 806, in save
    force_update=force_update, update_fields=update_fields)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/db/models/base.py", line 836, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/db/models/base.py", line 922, in _save_table
    result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/db/models/base.py", line 961, in _do_insert
    using=using, raw=raw)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/db/models/query.py", line 1063, in _insert
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 1099, in execute_sql
    cursor.execute(sql, params)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 80, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/srv/zulip/zerver/lib/db.py", line 32, in execute
    return wrapper_execute(self, super(TimeTrackingCursor, self).execute, query, vars)
  File "/srv/zulip/zerver/lib/db.py", line 19, in wrapper_execute
    return action(sql, params)
django.db.utils.ProgrammingError: relation "confirmation_invitekey" does not exist
LINE 1: INSERT INTO "confirmation_invitekey" ("invite_key", "date_cr...
                    ^
```
Should I be doing something else as well?

adnrs96 commented:
hmm, sorry I guess I need to run migrations.

hackerkid commented:
@adnrs96 I will update the comment to include run migrations. Thanks for reporting :)

hackerkid commented:
For testing 
```
 ./manage.py generate_multi_user_invite_link --realm zulip

 ./manage.py generate_multi_user_invite_link --realm zulip --limit 3

```
@rishig Thoughts?

rishig commented on zproject/urls.py:
```diff
@@ -136,6 +136,9 @@
     url(r'^register/(?P<realm_str>\S+)/$', zerver.views.registration.accounts_home_with_realm_str,
         name='zerver.views.registration.accounts_home_with_realm_str'),
 
+    url(r'^invite_users/(?P<confirmation_key>\S+)/$', zerver.views.registration.accounts_home_with_invite_key,
```
Let's do /join/<confirmation_key>

zulipbot-test commented:
Hello @hackerkid, a Zulip maintainer reviewed this pull request over a week ago, but you haven't updated your pull request since. Please take a look at the requested changes and update your pull request accordingly.

Thank you for your valuable contributions to Zulip!

rishig commented on zerver/tests/test_signup.py:
```diff
@@ -749,6 +749,68 @@ def test_if_emails_in_mixed_style_are_parsed_correctly(self):
+        result = self.client_post(invite_link, {'email': email})
+        self.assertEqual(result.status_code, 200)
+        self.assert_in_response("Whoops. The confirmation link is malformed.", result)
 
```
The tests I would do for this are:
* multiuse invite created 2 days ago works
* multiuse invite created 60 days ago does not work
* open realm with an expired multiuse invite should work (I think currently it doesn't?)
* create multiuse invite link with streams that aren't the default streams, and check that that works (I don't see where in the code that's currently being handled?)

hackerkid commented:
@rishig Updated the PR.
Some of the changes I made in this revision
*  Changed the model name to MultUseInvite invite.
* Added tests as you suggested
* Removed the mechanism of checking the number of times the link has used so far from the registration view.
* Implemented mechanism to subscribe users to specified streams.
*  multisuser_invite replaced with multiuse_invite and similar.

rishig commented on zerver/forms.py:
```diff
@@ -130,6 +132,7 @@ def clean_email(self):
 
         # Otherwise, the user is trying to join a specific realm.
         realm = self.realm
+        multiuse_invited = self.multiuse_invited
```
let's call this from_multiuse_invite

rishig commented on zerver/views/registration.py:
```diff
@@ -299,6 +300,17 @@ def accounts_home_with_realm_str(request, realm_str):
+        except ConfirmationKeyException as exception:
+            return render_confirmation_key_error(request, exception)
+
+        request.session["multiuse_invited"] = True
```
hm, it would be better to just process the confirmation key once, in this function, rather than passing it around.
Can we pass the multiuse_invite_object to accounts_home?

rishig commented on zerver/views/registration.py:
```diff
@@ -299,6 +300,17 @@ def accounts_home_with_realm_str(request, realm_str):
     else:
         return HttpResponseRedirect(reverse('zerver.views.registration.accounts_home'))
 
+def accounts_home_with_invite_key(request, confirmation_key):
```
I think `accounts_home_from_multiuse_invite` might be clearer

hackerkid commented:
> What happens if a stream is deleted in between when the invitation is sent and when they go through the registration process?

We won't show any error or warning in case stream gets deleted. User would get subscribed to one less stream. This is similar to the case of normal user invite. 

I have also updated the code :)

rishig commented:
oh sorry, I meant pass it as an argument to the accounts_home function, so that we didn't have to store it in the session. E.g. `accounts_home(request, multiuse_invite_object)`.

hackerkid commented:
> oh sorry, I meant pass it as an argument to the accounts_home function, so that we don't have to store it in the session. E.g. accounts_home(request, multiuse_invite_object).

I should not have used sessions here in first place. Updated. thanks :)

hackerkid commented:
@rishig Fixed the merge conflicts :)

hackerkid commented on zerver/tests/test_signup.py:
```diff
@@ -750,6 +748,117 @@ def test_if_emails_in_mixed_style_are_parsed_correctly(self):
+        # type: (List[Stream], Optional[datetime.datetime]) -> Text
+        invite = MultiuseInvite(realm=self.realm, referred_by=self.example_user("iago"))
+        invite.save()
+        if streams is not None:
```
`streams` has the type `ManyToManyField`. So both instance of `MultiuseInvite` and `Stream` should exist in database to do this operation. 

https://stackoverflow.com/questions/10366045/django-how-to-save-data-to-manytomanyfield

hackerkid commented on zerver/tests/test_signup.py:
```diff
@@ -750,6 +748,117 @@ def test_if_emails_in_mixed_style_are_parsed_correctly(self):
+        streams = [get_stream(stream_name, self.realm) for stream_name in stream_names]
+        invite_link = self.generate_multiuse_invite_link(streams=streams)
+        self.check_user_able_to_register(email2, invite_link)
+        self.check_user_subscribed_to_streams(name2, streams)
```
> I'd also check that the user is not subscribed to the default streams.

Actually the function make sure that the user is only subscribed to the streams he invited to and not anything else. Updated the function name to reflect this.

hackerkid commented:
@rishig I have fixed the issues and merge conflict. :)

rishig commented:
cool, this is looking good! I do want to get this into master soon, but it should wait for after the realms_have_subdomains work is merged, since it will simplify this code, and it is likely to create more merge conflicts. 

That being said, I think the next part can proceed in parallel, which is adding a checkbox to the invitation form, which says "Allow invitation link to be used multiple times [Learn more](https://url)", where we'll link to the appropriate /help document.

timabbott commented:
This seems clean enough and has tests, so I merged all but the last commit, after fixing a few minor style issues are renumbering the migration.  There are 2 priority follow-up projects:
* I think we should modify the management command to take the list of streams to use as an argument, and then we can merge that, too, which I think will be handy.
* I think in many cases, folks will just want the multi-use invite link to use the default streams of the organization, i.e. `streams=None`, rather than specifying them in advance, since you can update the default streams for the organization with the UI, and you can't update this through the UI.  I think this code doesn't actually support that, and if you try, you'll get `streams=[]`.  We need to fix that.

Once that's done, we should create one of these links and test it on chat.zulip.org.  

Since we want to use this feature for the GSoC Mentor Summit (and the email with this link should go in the next few days!), this should be your top priority @hackerkid.

hackerkid commented:
@timabbott Updated the management command.  The code we merged will sign up the users to default streams if no streams were mentioned in the invite.

timabbott commented:
Merged, thanks @hackerkid!
